src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,427 @@
+     (if (not= path new-path)
+       new-path
+       not-found))))
+
+
+
+(comment
+
+  '())
+
+
+;;============================================================================
+
+(>defn get-in
+  "Just like clojure.core/get-in, but if an element of the path is an ident it will follow the ident instead."
+  ([state-map path]
+   [map? vector? => any?]
+   (clojure.core/get-in state-map (tree-path->db-path state-map path)))
+
+  ([state-map path not-found]
+   [map? vector? any? => any?]
+   (clojure.core/get-in state-map (tree-path->db-path state-map path) not-found)))
+
+(comment
+
+  '())
+
+
+
+;;============================================================================
+
+
+;; TODO: see deep-remove-entity* (from incubator). I think the incubator implementation is technically insufficient. GC
+;; TODO should check/clean all tables, but only one level deep (not a walk). Also, the cascading would be useful to implement.
+(>defn remove-entity*
+  "Remove the given entity at the given ident. Also scans all tables and removes any to-one or to-many idents that are
+   found that match `ident` (removes dangling pointers to the removed entity).
+
+   The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+   (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+   Returns the new state map with the entity(ies) removed."
+
+  ;;TODO
+  ([state-map ident]
+   [map? eql/ident? => map?]
+   )
+
+  ;;TODO
+  ([state-map ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+   )
+  )
+
+
+
+(defn- dissoc-in
+  "Remove the given leaf of the `path` from recursive data structure `m`"
+  [m path]
+  (cond-> m
+          (clojure.core/get-in m (butlast path))
+          (update-in (butlast path) dissoc)))
+
+(defn- extract-all-idents
+  "Extracts all idents present in a children of a given entity"
+  [items]
+  (into []
+        (comp (keep (fn [v]
+                      (cond
+                        (eql/ident? v)
+                        [v]
+
+                        (and (vector? v)
+                             (every? eql/ident? v))
+                        v)))
+              cat)
+        (vals items)))
+
+
+(defn- dissoc-ident-from-top-level [state-map ident]
+  (cond-> state-map
+          (clojure.core/get-in state-map (butlast ident))
+          (update-in (butlast ident) dissoc (second ident))))
+
+(comment
+  (dissoc-ident-from-top-level normalized-state [:person/id 1])
+
+  '())
+
+
+;; TODO remove the dangling pointers to an ident
+(defn- dissoc-ident-from-nested-tables [state-map ident]
+  (cond-> state-map
+          (clojure.core/get-in state-map (butlast ident))
+          (update-in (butlast ident) dissoc (second ident))))
+
+(comment
+  (dissoc-ident-from-nested-tables normalized-state [:person/id 1])
+
+  '())
+
+
+
+
+;; REMOVE
+(defn deep-remove-entity*
```

Comment:
This one runs into stack overflow so this was the reason I've started to chunk this into different functions.

Reply:
Recursion is not necessary. There are only a few cases in a normalized db:

- There is a top level prop that points to the ident you care about...dissoc that entry
- There is a top-level props that is to-many, and includes the target ident. Remove the ident from that list.
- There is an attribute in some table entity that matches one of the prior two cases. Handle it just like you did for that, but for the entity.

You could write the first two cases as a single helper that is given a map and a key, and returns the map with that key properly changed (or dissoc'd). Then you could run that against the top-level map, and then use reduce to do the same against every table-looking thing in the db (top-level keys that point to a map whose values are all maps.